### DISCUSSION TITLE
error reading service account token - open : no such file or directory

### DISCUSSION THREAD

**Discussion by chrism-hutch** [kedacore/keda#3754] [General]
I'm trying to use Azure Workload, with Azure Queue as a trigger

```
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: crates
  namespace: default
spec:
  podIdentity:
    provider: azure-workload
    identityId: 0f71ba26-87c8-4b25-a9a7-e5085fbe9c3a
```
```
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: crates
  namespace: default
spec:
  scaleTargetRef:
    name: crates
  triggers:
  - type: azure-queue
    metadata:
      # Required
      queueName: somequeue
      # Optional, required when pod identity is used
      accountName: somehwere

      # Optional
      queueLength: "5" # default 5
    authenticationRef:
        name: crates
```
And this what the oprator logs, I've created my federated credentials which are tied to `0f71ba26-87c8-4b25-a9a7-e5085fbe9c3a`:
```
2022-10-17T12:35:34Z   INFO    Reconciling ScaledObject        {"controller": "scaledobject", "controllerGroup": "keda.sh", "controllerKind": "ScaledObject", "scaledObject": {"name":"crate│
│2022-10-17T12:36:04Z   ERROR   azure_queue_scaler      error)  {"type": "ScaledObject", "namespace": "default", "name": "cratedroprates", "error": "error reading service account token - open : │
│github.com/kedacore/keda/v2/pkg/scaling/cache.(*ScalersCache).IsScaledObjectActive                                                                                                            │
│       /workspace/pkg/scaling/cache/scalers_cache.go:89                                                                                                                                       │
│github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers                                                                                                                          │
│       /workspace/pkg/scaling/scale_handler.go:278                                                                                                                                            │
│github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop                                                                                                                        │
│       /workspace/pkg/scaling/scale_handler.go:149
```

Not sure what to think of this error, am I misusing Keda?

**Comment by chrism-hutch** [comment]
Alright, enabling workload (`--enable-workload-identity`) on my AKS cluster solved this

**Reply by remyletient** [reply]
It seems this option is no longer available in the latest version of azure-cli: https://learn.microsoft.com/en-us/cli/azure/aks?view=azure-cli-latest#az-aks-update

**Comment by piroinno** [comment]
Did not work for me sadly. Can you confirm if you installed keda with a service account

**Reply by chrism-hutch** [reply]
What error are you getting @piroinno? No turns out you need to enable workload identity when you deploy the helm chart as well, in the values, example:
```
helm install keda kedacore/keda --namespace keda --set podIdentity.azureWorkload.enabled=true,podIdentity.azureWorkload.tenantId=some-guid,podIdentity.azureWorkload.clientId=some-guid --create-namespace
```

**Reply by JorTurFer** [reply]
@piroinno, have you created the federation , updated KEDA with the needed values and you have workload-identity-hook deployed in your cluster?

**Reply by remyletient** [reply]
Same issue here.
The installation of azure-workload-identify and sample[ "quick-start" sample](https://azure.github.io/azure-workload-identity/docs/quick-start.html) work for me but when I try to configure `ScaledObject` and `TriggerAuthentification` like that that not work.


```
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: ia-bus-auth
  namespace: middleware
spec:
  podIdentity:
    provider: azure-workload
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
 name: classifier-python-scaled-by-metrics
 namespace: middleware
spec:
 minReplicaCount: 1                              # Optional. Default: 0
 maxReplicaCount: 1
 triggers:
   - type: azure-servicebus
     metadata:      
       namespace: dev-mw-bus
       topicName: topic_classifier_py
       subscriptionName: subs_request
       messageCount: "1"
     authenticationRef:
       name: ia-bus-auth
```
I've the followind error: 
```
Events:
  Type     Reason              Age                  From           Message
  ----     ------              ----                 ----           -------
  Normal   KEDAScalersStarted  29m                  keda-operator  Started scalers watch
  Normal   ScaledObjectReady   29m (x2 over 29m)    keda-operator  ScaledObject is ready for scaling
  Warning  KEDAScalerFailed    4m9s (x51 over 29m)  keda-operator  ChainedTokenCredential: failed to acquire a token.
Attempted credentials:
  AzureCLICredential: fork/exec /bin/sh: no such file or directory
  error reading service account token - open : no such file or directory
```
For my point of view a key difference is the explicit declaration of the service account name `workload-identity-sa` in the Pod spec for the `quick-start` sample. With the keda custom object I'm not able to specify such service account name, and that use the `default` that is not configured with right label.

Have you an idead ? Thanks ?

Keda version: 2.9.0
Kubernetes version: AKS 1.24.6

Regards,

**Reply by JorTurFer** [reply]
Using WI, you should bind KEDA service account with your managed identity. If you are using helm to deploy KEDA, [there is a section you can use to configure KEDA service account](https://github.com/kedacore/charts/blob/main/keda/values.yaml#L106-L123). If you are using kubectl with the manifest directly, you will need to modify them to add the required annotations to KEDA service account.
The error you are sending looks like the workload identity hook is not adding the needed resources to KEDA pod

**Reply by JorTurFer** [reply]
We have [an example using KEDA and workload identity here](https://github.com/kedacore/sample-dotnet-worker-servicebus-queue/blob/main/workload-identity.md), you can use it as a guide to configure KEDA properly

**Reply by remyletient** [reply]
It's ok now that works.

The points I've missed:
- the following option when install keda: podIdentity ...
```
> helm install keda kedacore/keda \
      --set podIdentity.azureWorkload.enabled=true \
      --set podIdentity.azureWorkload.clientId={app-id-for-keda-azure-ad-app} \
      --set podIdentity.azureWorkload.tenantId={tenant-id-for-keda-azure-ad-app} \
      --namespace keda-system
```
because I started from an existing Keda installation that use kubernetes secret to store the connectionString

- Create the federation credential:
```
cat <<EOF > params.json
{
  "name": "kubernetes-federated-credential",
  "issuer": "${SERVICE_ACCOUNT_ISSUER}",
  "subject": "system:serviceaccount:${SERVICE_ACCOUNT_NAMESPACE}:${SERVICE_ACCOUNT_NAME}",
  "description": "Kubernetes service account federated credential",
  "audiences": [
    "api://AzureADTokenExchange"
  ]
}
EOF

az ad app federated-credential create --id ${APPLICATION_OBJECT_ID} --parameters @params.json
```

Thanks you for you quick reaction, links ...

**Reply by JorTurFer** [reply]
you're welcome :) 
Happy to help!

**Reply by rtarasov** [reply]
Hi @remyletient ! Sorry for asking may be a stupid question but could please explain which SA should I use for federated-credentials? from which NS? Is it existing KEDA operator SA? or may be the new SA will be created by the command or need to be created in advance?

**Reply by JorTurFer** [reply]
you should use KEDA service account, all KEDA instances use the same SA (which is created during the deployment)

**Reply by rtarasov** [reply]
Thank you Jorge @JorTurFer for the answer! Regarding SA is clear now. But I'm a little bit confused about these two values:
      --set podIdentity.azureWorkload.clientId={app-id-for-keda-azure-ad-app} \
      --set podIdentity.azureWorkload.tenantId={tenant-id-for-keda-azure-ad-app} \
 We are installing the keda using helm through terraform and we have no any keda app in AD. Instead we have a managed identity for keda. Where I supposed to get these values from? Managed identity?
 Sorry for may be improper question and thank you in advance!

**Reply by rtarasov** [reply]
Answering to my own question - yes I should use the client id and tenant id from managed identity.

**Reply by JorTurFer** [reply]
yeah, you should use MSI values

**Reply by huzefaqubbawala** [reply]
@JorTurFer 
Is there any way to use this without helm. I have created service account using yaml file and created federated credentials as well. Now I am not sure how to perform the below step without using helm
--set podIdentity.azureWorkload.clientId={app-id-for-keda-azure-ad-app}
--set podIdentity.azureWorkload.tenantId={tenant-id-for-keda-azure-ad-app}

I am more comfortable with kubectl and yaml files, can anyone explain if this can be done using kubectl. I have installed KEDA using the below command

kubectl apply -f https://github.com/kedacore/keda/releases/download/v2.8.1/keda-2.8.1.yaml

**Reply by JorTurFer** [reply]
hi @huzefaqubbawala ,
You can see how to do it with kubectl in [workload identity docs](https://azure.github.io/azure-workload-identity/docs/quick-start.html#5-create-a-kubernetes-service-account)

**Reply by huzefaqubbawala** [reply]
Thanks @JorTurFer  for the response.

The link only talks about creating service account which I have already done before using kubectl. For now, I have gone ahead and used helm in my cluster. After using helm I faced the below error -

ChainedTokenCredential authentication failed POST https://login.microsoftonline.com/<tenantId>/oauth2/v2.0/token -------------------------------------------------------------------------------- RESPONSE 400 Bad Request -------------------------------------------------------------------------------- {   "error": "invalid_request",   "error_description": "AADSTS70021: No matching federated identity record found for presented assertion. Assertion Issuer: '<aks cluster issuer URL>'. Assertion Subject: 'system:serviceaccount:default:keda-operator'.

Background - 

I had already created this service account below -

apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    azure.workload.identity/client-id: <client Id>
  labels:
    azure.workload.identity/use: "true"
  name: workload-identity-sa
  namespace: default
-----------------------------------------------------
Also linked this to federated identity with below command -
az identity federated-credential create --name <name> --identity-name <user identity> --resource-group <rg> --issuer "<cluster issuer URL>" --subject system:serviceaccount:default:workload-identity-sa


From the error it looks like KEDA is using keda-operator service account to get token , Not sure how to tell KEDA to use the service account which I have created. To resolve the error , I have provided keda-operator federated identity.

**Reply by JorTurFer** [reply]
You should annotate KEDA service account, not yours. I mean, deploying KEDA, a new service account is created for it.
You have 2 option, federating (and annotating) the service account that is created or modifying the deployments to use your service account instead of keda-operator service account.
Using workload identity, you will use the pod service account, if you want to change it, you need to change it in the pod (at deployment level in this case).
Be care if you use your own service account instead of keda-operator, because keda-operator sa already has the required permissions that KEDA needs for working, if you change it you need to ensure that those permission are assigned to the sa

**Reply by huzefaqubbawala** [reply]
Thanks @JorTurFer .

This makes sense to me. 

So the federated identity should be created for keda-operator service account created by KEDA when it is installed by running below command -

az identity federated-credential create --name "name" --identity-name "identityName" --resource-group rgName --issuer "issuer url" --subject system:serviceaccount:default:keda-operator

This also means KEDA will always create service account using this name "keda-operator"

**Reply by JorTurFer** [reply]
if you use helm, the name is [a user given ](https://github.com/kedacore/charts/blob/main/keda/values.yaml#L131) with "keda-operator" as default. If you use the yamls, [it's always "keda-operator"](https://github.com/kedacore/keda/blob/main/config/service_account/service_account.yaml#L8)

**Reply by huzefaqubbawala** [reply]
Thank you so much @JorTurFer . This is super helpful

**Reply by JorTurFer** [reply]
yw!

**Reply by stephenleecariad** [reply]
I ran into the original error and was only able to get around it by patching the keda operator deployment and adding the 'azure.workload.identity/use: "true"' label.

ex: KedaCorePatch.yml:
spec:
  template:
    metadata:
      labels:
        azure.workload.identity/use: "true"
        
deploy step:
      - task: Kubernetes@1
        displayName: Patch Keda Deployment
        inputs:
          connectionType: "Azure Resource Manager"
          azureSubscriptionEndpoint: "xxxxxxxxxxxxxxxxxx"
          azureResourceGroup: "$(RESOURCE_GROUP_NAME)"
          kubernetesCluster: "$(RESOURCE_GROUP_NAME)-aks"
          useClusterAdmin: true
          command: patch
          arguments: deployment keda-operator -n keda --patch-file ./KedaCorePatch.yml

**Reply by JorTurFer** [reply]
Hey,
That's because they changed the required labels for being injected in the workload identity webhook, but we update them too in KEDA chart, if you use latest versions of KEDA using the helm chart, your patch is already included

**Reply by stephenleecariad** [reply]
Ah that would explain my problems then, I wasn't using a helm chart due to our company's reluctance to embrace it. I'll have to look into that!

**Reply by JorTurFer** [reply]
You aren't forced to use the chart at all, but you need to patch the required resources from your side. The best part of the chart here is that we update is with required stuff, but it's also true that it's another dep (I'm talking about helm)
